### ISSUE TITLE
Add documentation to use a Task.Supervisor

### ISSUE THREAD

**Issue by nickjj** [swoosh/swoosh#236]
Your README includes instructions on how to use `Task.start` for the simple case, but it would be very beneficial if you could include how to also implement a proper supervisor into an existing app and how to tie that into a `deliver_later` function.

**Comment by princemaple** [comment]
It's not much different from using any other library with Task.

- Start a `Task.Supervisor` with a name in your app's supervision tree
- Instead of calling `Task.start`, call `Task.Supervisor.start_child(sup_name, deliver_fun, options)`
- Done

As we explain below the example, sending email is very application specific. We don't want to make too many assumptions. And I don't think more examples are needed, since the use of `Task` module is so common, that it doesn't need to be taught in Swoosh README again.

What do you think?

**Comment by danschultzer** [comment]
I agree with you @princemaple, but you could also argue that `deliver_later` is something many developers are looking to implement (especially for developers who has moved from rails), and instructions on how to properly set up a supervisor could help save time for many. I don't think it belongs in the readme though, but maybe it would be good to have as a wiki page?

**Comment by princemaple** [comment]
@danschultzer thanks for pitching in. `Task.start` already covers `deliver_later`. How to setup a supervisor though, is useless if the developers don't understand what a supervisor is good for, why it's needed, and how to use it. If, on the other hand, they do understand the aforementioned points, they already don't need the guidance. That being said, I'm happy to setup a wiki page if anyone is willing to write up a short guide. Otherwise, we already recommend reading on the `Task` documentation. This is the real guide new developers should read :)

**Comment by nickjj** [comment]
Hi,

I mainly opened this because @stevedomin requested it (we were talking on Slack).

Basically, as a newcomer to Elixir, setting up a Task Supervisor in a custom app is a daunting task. But other email solutions like Bamboo have a `deliver_later` function that works out of the box and it's backed by a supervisor with no additional set up required by end users.

Demonstrating how to do that with documentation (or even implementing `deliver_later` at the library level) would be a pretty big win. Otherwise I think most people looking for email solutions and happen to be just getting into Elixir will end up using Bamboo.

**Comment by princemaple** [comment]
First of all, https://github.com/thoughtbot/bamboo/issues/126 (nothing has changed since)
We can totally add a suboptimal solution just like what bamboo has.
However, if a developer is blindly using `deliver_later` (at the level of implementation same as what bamboo has) it's not much better than simply using what we have in the docs. `Task.start`.

I know having an existing solution in the library can be very beneficial to new comers, but implementing something not much better than what bamboo has right now is not remotely useful in the long run. Ecto deprecated and removed callbacks from Schema (back then it was still called Model), so developers could correctly deal with data, with proper contextual and logical modules. It didn't have to, plenty of new devs ask about callbacks everyday coming from Rails. I personally use Swoosh in the way that:
- if I don't care about the email, I use `Task.start`
- if I do care, I use `Task.async` and other supervisor alternatives, so I can properly handle error and timeouts, etc

It doesn't take long to surpass the point where you rely on the library providing a `deliver_later` option. Take 30 minutes to read the docs on Task and Task.Supervisor, and maybe another hour about OTP, you won't need it anymore. You may even dislike having it.

Anyway, these are my opinions. Still, I'm open to adding a wiki page / readme section explaining this kind of things. And if Steve wants to implement something simple, I'm not against it either.

**Comment by nickjj** [comment]
Yeah with ActiveJob you would typically use something like Sidekiq to somewhat intelligently handle failures and retries, or at least get some type of insight on what flowed through your queue.

I was hoping to avoid that type of model, but it's sounding like if you to do want retries, etc. then something comparable to task_bunny is likely necessary, short of writing your own ETS backed solution?

A repo or wiki example that walks through the "real" way to make a robust solution would be awesome. I know it's not directly your responsibility, but email is usually the first thing a web developer will encounter where they want to reach for something happening asynchronously.

**Comment by princemaple** [comment]
@nickjj No I don't use task_bunny. Elixir itself is good enough for me. I don't have a need for queuing.
Most of the time I send asynchronously to avoid blocking other operations (e.g. responding to a request), but the emails are almost always sent immediately at the same time (in parallel to other things that need to happen without being blocked).

**Comment by princemaple** [comment]
When I do have to handle complicated failures, I wrap the entire asynchronous part with a GenServer, so I can start a dedicated server to wait on the email being sent, and handle all possible outcomes.

**Comment by princemaple** [comment]
I made this little lib that deals with many aspects of running async tasks, which I described in my previous comments. https://github.com/princemaple/tusk

**Comment by paveltyk** [comment]
How do I assert email sent in tests, when using a separate task? `assert_email_sent` looks in current process mailbox. I've tried:

```
:erlang.trace(email_pid, true, [:receive])
assert_receive {:trace, ^email_pid, :receive, {:email, _}}
```

But no luck. Any advice?

**Comment by princemaple** [comment]
hi @paveltyk saw your question while sweeping old issues. If you are still wondering, please take a look at https://github.com/swoosh/swoosh/issues/425
